File: 1396-design-underground-system/1396-design-underground-system.py
class UndergroundSystem:

    def __init__(self):
        self.customerCheckIn = collections.defaultdict(tuple)
        self.train = collections.defaultdict(tuple)
    

    def checkIn(self, id: int, stationName: str, t: int) -> None:
        self.customerCheckIn[id] = (stationName, t)

    def checkOut(self, id: int, checkOutStation: str, t: int) -> None:
        checkInStation, checkInTime = self.customerCheckIn[id]
        del self.customerCheckIn[id]
        
        if (checkInStation, checkOutStation) not in self.train:
            self.train[(checkInStation, checkOutStation)] = (t-checkInTime ,1)
            return
        
        prevAvgTime, prevCount = self.train[(checkInStation, checkOutStation)]
        self.train[(checkInStation, checkOutStation)] = ( (prevAvgTime*(prevCount)+(t-checkInTime))/(prevCount+1), prevCount+1)
        return
    
    def getAverageTime(self, startStation: str, endStation: str) -> float:
        return self.train[(startStation, endStation)][0]


# Your UndergroundSystem object will be instantiated and called as such:
# obj = UndergroundSystem()
# obj.checkIn(id,stationName,t)
# obj.checkOut(id,stationName,t)
# param_3 = obj.getAverageTime(startStation,endStation)

File: 1396-design-underground-system/1396-design-underground-system.java
class UndergroundSystem {
    Map<Integer, Pair<String, Integer>> checkInMap = new HashMap<>();
    Map<String, Pair<Double, Integer>> routeMap = new HashMap<>();
    
    
    public UndergroundSystem() {
        
    }
    
    public void checkIn(int id, String stationName, int t) {
        checkInMap.put(id, new Pair<>(stationName, t));
    }
    
    public void checkOut(int id, String stationName, int t) {
        Pair<String, Integer> checkInInfo = checkInMap.get(id);
        checkInMap.remove(id);
        
        String routeName = checkInInfo.getKey() + "_" + stationName;
        Integer checkInTime = checkInInfo.getValue();
        
        Pair<Double, Integer> oldRoute = routeMap.getOrDefault(routeName, new Pair<>(0.0, 0));
        routeMap.put(routeName, new Pair<> ( (oldRoute.getKey()*oldRoute.getValue()+(t-checkInTime))/(oldRoute.getValue()+1), oldRoute.getValue()+1) );
        
        
    }
    
    public double getAverageTime(String startStation, String endStation) {
        return routeMap.get(new String (startStation + "_" + endStation) ).getKey();
    }
}

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * UndergroundSystem obj = new UndergroundSystem();
 * obj.checkIn(id,stationName,t);
 * obj.checkOut(id,stationName,t);
 * double param_3 = obj.getAverageTime(startStation,endStation);
 */